crates/iceberg/src/expr/visitors/manifest_evaluator.rs

```diff
@@ -98,103 +101,249 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
+            let Literal::Primitive(PrimitiveLiteral::String(upper_bound)) = upper_bound else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform starts_with on non-string upper bound",
+                ));
+            };
+
+            let min_len = upper_bound.chars().count().min(prefix_len);
+            let truncated_upper_bound = String::from(&upper_bound[..min_len]);
+            if prefix > &truncated_upper_bound {
+                return ROWS_CANNOT_MATCH;
+            }
+        }
+
+        ROWS_MIGHT_MATCH
     }
 
     fn not_starts_with(
         &mut self,
         reference: &BoundReference,
-        literal: &Datum,
+        datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+
+        if field.contains_null || field.lower_bound.is_none() || field.upper_bound.is_none() {
+            return ROWS_MIGHT_MATCH;
+        }
+
+        let PrimitiveLiteral::String(prefix) = datum.literal() else {
+            return Err(Error::new(
+                ErrorKind::Unexpected,
+                "Cannot perform not_starts_with on non-string value",
+            ));
+        };
+
+        let prefix_len = prefix.chars().count();
+
+        // not_starts_with will match unless all values must start with the prefix. This happens when
+        // the lower and upper bounds both start with the prefix.
+        if let Some(lower_bound) = &field.lower_bound {
+            let Literal::Primitive(PrimitiveLiteral::String(lower_bound)) = lower_bound else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform not_starts_with on non-string lower bound",
+                ));
+            };
+
+            // if lower is shorter than the prefix then lower doesn't start with the prefix
+            if prefix_len > lower_bound.chars().count() {
+                return ROWS_MIGHT_MATCH;
+            }
+
+            let truncated_lower_bound = String::from(&lower_bound[..prefix_len]);
+            if prefix == &truncated_lower_bound {
+                if let Some(upper_bound) = &field.upper_bound {
+                    let Literal::Primitive(PrimitiveLiteral::String(upper_bound)) = upper_bound
+                    else {
+                        return Err(Error::new(
+                            ErrorKind::Unexpected,
+                            "Cannot perform not_starts_with on non-string upper bound",
+                        ));
+                    };
+
+                    // if upper is shorter than the prefix then upper can't start with the prefix
+                    if prefix_len > upper_bound.chars().count() {
+                        return ROWS_MIGHT_MATCH;
+                    }
+
+                    let truncated_upper_bound = String::from(&upper_bound[..prefix_len]);
```

Comment:
You can't safely use `prefix_len` to index into `upper_bound` as there is no guarantee that the index will fall on a valid UTF8 character boundary, and if it doesn't then Rust will panic.

See here for an example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1e4950cb298bd87560be74fa48576184

You need to do this instead:

```rust
let truncated_upper_bound = upper_bound.chars().take(prefix_len).collect::<String>();
```

(See here for working example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=86969f315ef1bdc4a0863bde9356c1f4)